### ISSUE TITLE
ConnectionError occasionally in CI could be fixed via exp backoff

### ISSUE THREAD

**Issue by mattjegan** [kennethreitz/pipenv#1113]
When using pipenv in a CI environment occasionally the provider's internet connection becomes unstable and this can cause `ConnectionError`s through pipenv. I propose that this could be handled better than just throwing a trace or logging the error by instead implementing a retry but doubling the time between tries up to some upper bound, say 16 seconds (4 tries).

##### Describe your environment

1. Alpine Linux
1. Python version: `3.6.0`
1. Pipenv version: `8.2.6`

##### Expected result

Upon failing some requests pipenv could've retried rather than throw a `ConnectionError`.

##### Actual result

```
.
.
.
  File "/usr/local/lib/python3.6/site-packages/pipenv/utils.py", line 495, in resolve_deps
    r = requests.get('https://pypi.org/pypi/{0}/json'.format(name), timeout=10)
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 521, in get
    return self.request('GET', url, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 508, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 618, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/adapters.py", line 490, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))
```

##### Steps to replicate

Unknown (`pipenv install`), because it's the peer (PyPI) that reset the connection.

**Comment by nateprewitt** [comment]
Thanks for notifying us @mattjegan. We're already tracking this in #1110 so I'm closing this out as a dupe.

**Comment by mattjegan** [comment]
@nateprewitt thanks for the quick reply, I did notice that issue but thought of this more as a new feature which would implement a retry rather than the fix that is proposed in #1110 which is just logging an error message and aborting still. Would you prefer I follow up there?

**Comment by nateprewitt** [comment]
@mattjegan the proposal in #1110 isn't to abort but rather skip the optional step of retrieving meta data and moving on with the installation instead of failing.

Either way though, it's much easier to keep track of issues in one ticket. You're welcome to propose the exponential back off solution there and we can discuss trade offs.

I will say though from initial consideration, that may be slightly better if it's a transient network failure but has a considerably worse worst case if it's a persistent network issue. We'll continue to retry n times for every package. Probably better to fail fast and move onto to the next steps.
